hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/bucket/TestBucketCachePersister.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.hadoop.hbase.io.hfile.HFileContext;
+import org.apache.hadoop.hbase.io.hfile.HFileContextBuilder;
+import org.apache.hadoop.hbase.io.hfile.RandomKeyValueUtil;
+import org.apache.hadoop.hbase.regionserver.StoreFileWriter;
+import org.apache.hadoop.hbase.testclassification.IOTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.TestName;
+
+@Category({ IOTests.class, MediumTests.class })
+public class TestBucketCachePersister {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestBucketCachePersister.class);
+
+  public TestName name = new TestName();
+
+  public int constructedBlockSize = 16 * 1024;
+
+  public int[] constructedBlockSizes =
+    new int[] { 2 * 1024 + 1024, 4 * 1024 + 1024, 8 * 1024 + 1024, 16 * 1024 + 1024,
+      28 * 1024 + 1024, 32 * 1024 + 1024, 64 * 1024 + 1024, 96 * 1024 + 1024, 128 * 1024 + 1024 };
+
+  private static final HBaseTestingUtil TEST_UTIL = new HBaseTestingUtil();
+
+  private static final int NUM_VALID_KEY_TYPES = KeyValue.Type.values().length - 2;
+  private static final int DATA_BLOCK_SIZE = 2048;
+  private static final int NUM_KV = 1000;
+
+  final long capacitySize = 32 * 1024 * 1024;
+  final int writeThreads = BucketCache.DEFAULT_WRITER_THREADS;
+  final int writerQLen = BucketCache.DEFAULT_WRITER_QUEUE_ITEMS;
+
+  @Test
+  public void testPrefetchPersistenceCrash() throws Exception {
+
+    Configuration conf;
+    String prefetchPersistencePath;
+    Path testDir;
+
+    conf = TEST_UTIL.getConfiguration();
+    conf.setBoolean(CacheConfig.PREFETCH_BLOCKS_ON_OPEN_KEY, true);
+    testDir = TEST_UTIL.getDataTestDir();
+    TEST_UTIL.getTestFileSystem().mkdirs(testDir);
+    prefetchPersistencePath = testDir + "/prefetch.persistence";
+    conf.set(CacheConfig.PREFETCH_PERSISTENCE_PATH_KEY, prefetchPersistencePath);
+    FileSystem fs = HFileSystem.get(conf);
+    long BUCKETCACHE_PERSIST_INTERVAL = 3000;
+
+    conf.setLong(CacheConfig.BUCKETCACHE_PERSIST_INTERVAL_KEY, BUCKETCACHE_PERSIST_INTERVAL);
+    long startTime = System.currentTimeMillis();
+    BucketCache bucketCache = new BucketCache("file:" + testDir + "/bucket.cache", capacitySize,
+      constructedBlockSize, constructedBlockSizes, writeThreads, writerQLen,
+      testDir + "/bucket.persistence", 60 * 1000, conf);
+    CacheConfig cacheConf = new CacheConfig(conf, bucketCache);
+    // Load Cache
+    Path storeFile = writeStoreFile("TestPrefetch0", conf, cacheConf, fs);
+    Path storeFile2 = writeStoreFile("TestPrefetch1", conf, cacheConf, fs);
+    readStoreFile(storeFile, 0, fs, cacheConf, conf, bucketCache);
+    readStoreFile(storeFile2, 0, fs, cacheConf, conf, bucketCache);
+    Thread.sleep(BUCKETCACHE_PERSIST_INTERVAL - startTime + System.currentTimeMillis() + 100);
```

Comment:
So readStoreFile makes sure to only return once prefetch is complete. Do we still this extra (- startTime + System.currentTimeMillis()), plus the 100 milis to pass this test?